README.txt
This extension lets you extract metadata from (yet) pictures containing IPTC or EXIF data.

Example:

10 = FILEINFO
10 {
	#debug = 1
	
	file = fileadmin/myfile.jpg
	metadata.field = EXIF:Author // IPTC:2#080|0	
	
	wrap = <strong> | </strong>
}

This will first try to extract author from EXIF, and if not found, will read it from IPTC.
IPTC stores the photograph in 2#080, subarray 0, that is IPTC:2#080|0 with this extension.

You'll find a description of IPTC field on

http://ch2.php.net/manual/en/function.iptcparse.php

or both for EXIF and IPTC if you activate the debugging option in TS.

Enjoy!

ChangeLog
2015-02-17  Xavier Perseguers  <xavier@causal.ch>

	* [TASK] Raise version to 0.4.0 (obsolete)

2014-11-08  Xavier Perseguers  <xavier@causal.ch>

	* [BUGFIX] Prevent use of XMP toolkit when no metadata are available

2014-05-04  Xavier Perseguers  <xavier@causal.ch>

	* [TASK] Set dependency to TYPO3 4.5-4.7 (tested)
	* [BUGFIX] Invalid call to debug method

2010-10-20  Xavier Perseguers  <typo3@perseguers.ch>

	* Integrated library PHP_JPEG_Metadata_Toolkit to retrieve XMP metadata (PDF files)

2010-09-17  Xavier Perseguers  <typo3@perseguers.ch>

	* Fixed PHP Warning: Invalid argument supplied for foreach()

2010-08-22  Xavier Perseguers  <typo3@perseguers.ch>

	* Raised version to 0.3.0 (beta)
	* Updated documentation with additional examples

2010-08-16  Xavier Perseguers  <typo3@perseguers.ch>

	* Raised version to 0.2.0
	* Added geolocation (latitude/longitude/altitude) on sample picture wbtc05.jpg
	* Enhanced debugging information

2010-03-17  Xavier Perseguers  <typo3@perseguers.ch>

	* Added support for dynamically generated file name (thanks to Georg Ringer)
	* Updated documentation

2009-03-13  Xavier Perseguers  <typo3@perseguers.ch>

	* Added support for external tools

2008-07-17  Xavier Perseguers  <typo3@perseguers.ch>

	* Initial release


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "metadata_ts".
 *
 * Auto generated 17-02-2015 09:51
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Metadata in TS',
	'description' => 'Lets you extract metadata from different file formats using TypoScript.',
	'category' => 'misc',
	'author' => 'Xavier Perseguers (Causal)',
	'author_email' => 'xavier@causal.ch',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'obsolete',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.4.0',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.5.0-4.7.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:66:{s:9:"ChangeLog";s:4:"788e";s:21:"ext_conf_template.txt";s:4:"7d1d";s:12:"ext_icon.gif";s:4:"5807";s:17:"ext_localconf.php";s:4:"253c";s:10:"README.txt";s:4:"3935";s:32:"api/class.user_metadata_cobj.php";s:4:"4194";s:14:"doc/manual.sxw";s:4:"a1fc";s:41:"lib/PHP_JPEG_Metadata_Toolkit/COPYING.txt";s:4:"393a";s:48:"lib/PHP_JPEG_Metadata_Toolkit/Edit_File_Info.php";s:4:"dc98";s:56:"lib/PHP_JPEG_Metadata_Toolkit/Edit_File_Info_Example.php";s:4:"b16f";s:41:"lib/PHP_JPEG_Metadata_Toolkit/Example.php";s:4:"a948";s:38:"lib/PHP_JPEG_Metadata_Toolkit/EXIF.php";s:4:"3996";s:48:"lib/PHP_JPEG_Metadata_Toolkit/EXIF_Makernote.php";s:4:"e8cd";s:43:"lib/PHP_JPEG_Metadata_Toolkit/EXIF_Tags.php";s:4:"0354";s:49:"lib/PHP_JPEG_Metadata_Toolkit/get_casio_thumb.php";s:4:"27cf";s:48:"lib/PHP_JPEG_Metadata_Toolkit/get_exif_thumb.php";s:4:"daae";s:48:"lib/PHP_JPEG_Metadata_Toolkit/get_JFXX_thumb.php";s:4:"3a77";s:51:"lib/PHP_JPEG_Metadata_Toolkit/get_minolta_thumb.php";s:4:"c610";s:46:"lib/PHP_JPEG_Metadata_Toolkit/get_ps_thumb.php";s:4:"541e";s:38:"lib/PHP_JPEG_Metadata_Toolkit/IPTC.php";s:4:"7922";s:38:"lib/PHP_JPEG_Metadata_Toolkit/JFIF.php";s:4:"a209";s:38:"lib/PHP_JPEG_Metadata_Toolkit/JPEG.php";s:4:"2647";s:53:"lib/PHP_JPEG_Metadata_Toolkit/Photoshop_File_Info.php";s:4:"3eae";s:47:"lib/PHP_JPEG_Metadata_Toolkit/Photoshop_IRB.php";s:4:"5a0b";s:45:"lib/PHP_JPEG_Metadata_Toolkit/PictureInfo.php";s:4:"3ac8";s:37:"lib/PHP_JPEG_Metadata_Toolkit/PIM.php";s:4:"0e01";s:44:"lib/PHP_JPEG_Metadata_Toolkit/pjmt_utils.php";s:4:"18cb";s:38:"lib/PHP_JPEG_Metadata_Toolkit/test.jpg";s:4:"5b34";s:45:"lib/PHP_JPEG_Metadata_Toolkit/TIFFExample.php";s:4:"b185";s:49:"lib/PHP_JPEG_Metadata_Toolkit/Toolkit_Version.php";s:4:"bd03";s:41:"lib/PHP_JPEG_Metadata_Toolkit/Unicode.php";s:4:"4fd3";s:49:"lib/PHP_JPEG_Metadata_Toolkit/Write_File_Info.php";s:4:"1caa";s:37:"lib/PHP_JPEG_Metadata_Toolkit/XML.php";s:4:"d7d1";s:37:"lib/PHP_JPEG_Metadata_Toolkit/XMP.php";s:4:"2efe";s:49:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/agfa.php";s:4:"6491";s:50:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/canon.php";s:4:"ddba";s:50:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/casio.php";s:4:"9a9f";s:50:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/epson.php";s:4:"a378";s:53:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/fujifilm.php";s:4:"e9fb";s:59:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/konica_minolta.php";s:4:"2c01";s:52:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/kyocera.php";s:4:"9c8a";s:50:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/nikon.php";s:4:"2917";s:52:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/olympus.php";s:4:"c283";s:54:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/panasonic.php";s:4:"c5be";s:51:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/Pentax.php";s:4:"ed06";s:50:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/ricoh.php";s:4:"c2d0";s:49:"lib/PHP_JPEG_Metadata_Toolkit/Makernotes/sony.php";s:4:"5c0b";s:63:"lib/PHP_JPEG_Metadata_Toolkit/documentation/Camera_List_1.0.pdf";s:4:"d82d";s:56:"lib/PHP_JPEG_Metadata_Toolkit/documentation/changes.html";s:4:"bda3";s:58:"lib/PHP_JPEG_Metadata_Toolkit/documentation/css_terms.html";s:4:"ab40";s:69:"lib/PHP_JPEG_Metadata_Toolkit/documentation/edit_write_file_info.html";s:4:"885c";s:56:"lib/PHP_JPEG_Metadata_Toolkit/documentation/example.html";s:4:"f5f4";s:57:"lib/PHP_JPEG_Metadata_Toolkit/documentation/examples.html";s:4:"1f64";s:53:"lib/PHP_JPEG_Metadata_Toolkit/documentation/exif.html";s:4:"513b";s:54:"lib/PHP_JPEG_Metadata_Toolkit/documentation/index.html";s:4:"9c64";s:54:"lib/PHP_JPEG_Metadata_Toolkit/documentation/intro.html";s:4:"eadc";s:53:"lib/PHP_JPEG_Metadata_Toolkit/documentation/jfif.html";s:4:"a2ac";s:53:"lib/PHP_JPEG_Metadata_Toolkit/documentation/jpeg.html";s:4:"bb9a";s:58:"lib/PHP_JPEG_Metadata_Toolkit/documentation/photoshop.html";s:4:"14a8";s:68:"lib/PHP_JPEG_Metadata_Toolkit/documentation/photoshop_file_info.html";s:4:"a64e";s:61:"lib/PHP_JPEG_Metadata_Toolkit/documentation/picture_info.html";s:4:"734a";s:53:"lib/PHP_JPEG_Metadata_Toolkit/documentation/style.css";s:4:"9b8b";s:60:"lib/PHP_JPEG_Metadata_Toolkit/documentation/tiffexample.html";s:4:"6e11";s:53:"lib/PHP_JPEG_Metadata_Toolkit/documentation/todo.html";s:4:"8876";s:52:"lib/PHP_JPEG_Metadata_Toolkit/documentation/xmp.html";s:4:"4ccd";s:18:"samples/wbtc05.jpg";s:4:"52b1";}',
	'suggests' => array(
	),
);

?>

api/class.user_metadata_cobj.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2008-2014 Xavier Perseguers <xavier@causal.ch>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*  A copy is found in the textfile GPL.txt and important notices to the license
*  from the author is found in LICENSE.txt distributed with these scripts.
*
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/


/**
 * Handles TS content object FILEINFO.
 *
 * $Id$
 */
class user_metadata_cobj {

	protected $data = array();
	protected $conf = array();
	protected $has_exif = false;
	protected $has_iptc = false;

	/**
	 * Rendering the cObject, FILEINFO
	 *
	 * @param	string		$name: name of the cObject ('FILEINFO')
	 * @param	array		$conf: array of TypoScript properties
	 * @param	string		$TSkey: TS key set to this cObject
	 * @return	string		output
	 */
	function cObjGetSingleExt($name, $conf, $TSkey, &$oCObj) {
		$this->conf = $conf;

		$file = $oCObj->stdWrap($conf['file'], $conf['file.']);
		$field = $this->conf['metadata.']['field'];

		if (!t3lib_div::isAbsPath($file)) {
			$file = t3lib_div::getFileAbsFileName($file);
		}

		$fields = t3lib_div::trimExplode('//', $field, 1);
		if ($this->conf['debug'] && count($fields) == 0) {
				// Make sure both EXIF and IPTC are available
			$fields[] = 'EXIF:';
			$fields[] = 'IPTC:';
			$fields[] = 'XMP:';
		}

		foreach ($fields as $f) {
			list($type, $metatag) = t3lib_div::trimExplode(':', $f);

			switch ($type) {
				case 'EXIF': $this->EXIF($file); break;
				case 'IPTC': $this->IPTC($file); break;
				case 'XMP': $this->XMP($file); break;
			}
		}

		if ($this->conf['debug']) {
			t3lib_utility_Debug::debug($this->data);
		}

		$content = $this->getFieldVal($field);

		return $oCObj->stdWrap($content, $this->conf);
	}

	/**
	 * Reads EXIF metatags.
	 *
	 * @param string filename of the image
	 * @return void
	 */
	protected function EXIF($file) {
		if ($this->has_exif) return;

		$extconf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['metadata_ts']);
		$exif_data = array();

		if (file_exists($file)) {
			if ($extconf['exifTool']) {
				//$cmd = t3lib_exec::getCommand($extconf['exifTool']) . ' "' . $file . '"';
				$cmd = $extconf['exifTool'] . ' "' . $file . '"';
				exec($cmd, $exif, $ret = '');

				if (!$ret && is_array($exif)) {
					$exif_data = self::extractData($exif, $extconf['exifColumnSeparator'], $extconf['exifKeyColumn'], $extconf['exifValueColumn']);
				} else {
					// Debug this problem
				}

			} else {
				$image_info = getimagesize($file);

				if ($image_info[2] == 2) { // check for correct image-type
					$exif_data = exif_read_data($file, TRUE, FALSE); // Load all EXIF informations from the original Picture in an array
					$exif_data['Comments'] = htmlentities(str_replace("\n", '<br />', $exif_array['Comments'])); // Linebreak
					$exif_data = self::formatGpsData($exif_data);
				}
			}
		}

			// Put all data into $this->data
		$this->storeArray('EXIF:', '', $exif_data);
	}

	/**
	 * Reads IPTC metatags.
	 *
	 * @param string filename of the image
	 * @return void
	 */
	protected function IPTC($file) {
		if ($this->has_iptc) return;

		$extconf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['metadata_ts']);
		$iptc_data = array();

		if (file_exists($file)) {
			if ($extconf['iptcTool']) {
				//$cmd = t3lib_exec::getCommand($extconf['iptcTool']) . ' "' . $file . '"';
				$cmd = $extconf['iptcTool'] . ' "' . $file . '"';
				exec($cmd, $iptc, $ret = '');

				if (!$ret && is_array($iptc)) {
					$iptc_data = self::extractData($iptc, $extconf['iptcColumnSeparator'], $extconf['iptcKeyColumn'], $extconf['iptcValueColumn']);
				} else {
					// Debug this problem
				}

			} else {
				getimagesize($file, $image_info);

				if (is_array($image_info)) {
					$iptc_data = iptcparse($image_info['APP13']);
				}
			}


		}

		if (is_array($iptc_data)) {
			// Put all data into $this->data
			$this->storeArray('IPTC:', '', $iptc_data);
		}
	}

	/**
	 * Reads XMP metatags from a PDF file.
	 *
	 * @param string $file
	 * @return void
	 */
	protected function XMP($file) {
		if (strtolower(substr($file, -4)) !== '.pdf') return;

		$extconf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['metadata_ts']);
		$xmp_data = array();

		if (file_exists($file)) {
			if ($extconf['pdfinfoTool']) {
				//$cmd = t3lib_exec::getCommand($extconf['pdfinfoTool']) . ' -meta "' . $file . '"';
				$cmd = $extconf['pdfinfoTool'] . ' -meta "' . $file . '"';

				exec($cmd, $pdfmeta, $ret = '');
				if (!$ret && is_array($pdfmeta) && count($pdfmeta) > 0) {
					$xmlStr = '';
					$output = FALSE;
					foreach ($pdfmeta as $line) {
						if (t3lib_div::isFirstPartOfStr($line, '<x:xmpmeta')) {
							$output = TRUE;
						} elseif (t3lib_div::isFirstPartOfStr($line, '<'.'?xpacket')) {
							$output = FALSE;
						}
						if ($output) {
							$xmlStr .= $line;
						}
					}

					require_once(t3lib_extMgm::extPath('metadata_ts') . 'lib/PHP_JPEG_Metadata_Toolkit/XMP.php');

					$xmp = read_XMP_array_from_text($xmlStr);
					$xmp_data = Interpret_XMP_to_Array($xmp);
				}
			}
		}

		if ($xmp_data) {
			$this->storeArray('XMP:', '', $xmp_data);
		}
	}

	/**
	 * Extracts key/value pairs from a command-line output ($data)
	 *
	 * @param	array	$data
	 * @param	string	$separator
	 * @param	integer	$keyCol
	 * @param	integer	$valueCol
	 */
	protected static function extractData($data, $separator, $keyCol, $valueCol) {
		$info = array();
		foreach ($data as $content) {
			$separator = '\s+';
			$temp = preg_split("/$separator/", $content, $valueCol + 1);

			$key = $temp[$keyCol];
			$value = $temp[$valueCol];

			if (preg_match('/^lang=".*" (.*)$/', $value, $matches)) {
				$value = $matches[1];
			}

			$info[$key] = $value;
		}

		return $info;
	}

	/**
	 * Stores metadata array into protected data array.
	 *
	 * @param	string		$prefix
	 * @param	string		$key
	 * @param	array		$values
	 */
	function storeArray($prefix, $key, $values) {
		foreach ($values as $subkey => $value) {
			$datakey = '';
			if ($key) $datakey = $key . '|';
			$datakey .= $subkey;

			if (is_array($value)) {
				$this->storeArray($prefix, $datakey, $value);
			} else {
				$this->data[$prefix. $datakey] = $value;
			}
		}
	}

	/**
	 * Returns the value for the field from $this->data. If "//" is found in the $field value that token will split the field values apart and the first field having a non-blank value will be returned.
	 *
	 * @param	string		The fieldname, eg. "title" or "navtitle // title" (in the latter case the value of $this->data[navtitle] is returned if not blank, otherwise $this->data[title] will be)
	 * @return	string
	 */
	function getFieldVal($field) {
		if (!strstr($field, '//')) {
			return $this->data[trim($field)];
		} else {
			$sections = t3lib_div::trimExplode('//', $field, 1);
			foreach ($sections as $section) {
				if (strcmp($this->data[$section], '')) {
					return $this->data[$section];
				}
			}
		}
	}

	/**
	 * Formats GPS data.
	 *
	 * @param array $exifData
	 * @return array
	 */
	protected static function formatGpsData($exifData) {
		if (isset($exifData['GPSLatitude'])) {
			$latitude = $exifData['GPSLatitude'];
			$exifData['GPSLatitudePosition'] = ($exifData['GPSLatitudeRef'] !== 'N' ? '-' : '') . self::calcGpsPosition($latitude);
		}
		if (isset($exifData['GPSLongitude'])) {
			$longitude = $exifData['GPSLongitude'];
			$exifData['GPSLongitudePosition'] = ($exifData['GPSLongitudeRef'] !== 'E' ? '-' : '') . self::calcGpsPosition($longitude);
		}
		if (isset($exifData['GPSLatitudePosition']) && $exifData['GPSLongitudePosition']) {
			$exifData['GPSLink'] = sprintf('http://maps.google.com/maps?q=%s,%s', $exifData['GPSLatitudePosition'], $exifData['GPSLongitudePosition']);
		}
		return $exifData;
	}

	/**
	 * Calculates GPS position.
	 *
	 * @param array $gpsData
	 * @return double
	 */
	protected static function calcGpsPosition($gpsData) {
		if (function_exists('bcscale')) {
			bcscale(14);
			$pos = self::evalFloat($gpsData[0]);
			$pos = bcadd($pos, bcdiv(self::evalFloat($gpsData[1]), 60));
			$pos = bcadd($pos, bcdiv(self::evalFloat($gpsData[2]), 3600));
		} else {
			$pos = self::evalFloat($gpsData[0]);
			$pos += self::evalFloat($gpsData[1]) / 60;
			$pos += self::evalFloat($gpsData[2]) / 3600;
		}
		return $pos;
	}

	/**
	 * Evaluates a fractional value.
	 *
	 * @param $frac
	 * @return float
	 */
	protected static function evalFloat($frac) {
		$matches = array();
		if (preg_match('/(\d+)\/(\d+)/', $frac, $matches)) {
			return $matches[1] / $matches[2];
		} else {
			return 0;
		}
	}

}


if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/metadata_ts/api/class.user_metadata_cobj.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/metadata_ts/api/class.user_metadata_cobj.php']);
}
?>